core/src/main/java/org/elasticsearch/search/aggregations/metrics/stats/multifield/MultiFieldStatsResults.java

```diff
@@ -0,0 +1,431 @@
+
+            // update running sum
+            fieldSum.put(fieldName, fieldValue + (fieldSum.containsKey(fieldName) ? fieldSum.get(fieldName) : 0));
+
+            // update running deltas
+            delta.put(fieldName, fieldValue * docCount - fieldSum.get(fieldName));
+
+            // update running mean, variance, skewness, kurtosis
+            if (means.containsKey(fieldName) == true) {
+                // update running means
+                final double prevMean = means.get(fieldName);
+                final double d = fieldValue - prevMean;
+                final double dn = d / docCount;
+                final double dn2 = dn * dn;
+                final double term1 = d * dn * (docCount - 1);
+                final double m2 = variances.get(fieldName);
+                means.put(fieldName, prevMean + d / docCount);
+                variances.put(fieldName, variances.get(fieldName) + term1);
+                final double m3 = skewness.get(fieldName);
+                skewness.put(fieldName, m3 + (term1 * dn * (docCount - 2D) - 3D * dn * m2));
+                final double newKurt = term1 * dn2 * (docCount * docCount - 3D * docCount + 3D) + 6D * dn2 * m2 - 4D * dn * m3;
+                kurtosis.put(fieldName, kurtosis.get(fieldName) + newKurt);
+            } else {
+                means.put(fieldName, fieldValue);
+                variances.put(fieldName, 0.0);
+                skewness.put(fieldName, 0.0);
+                kurtosis.put(fieldName, 0.0);
+            }
+        }
+
+        // deep copy of hash keys (field names)
+        ArrayList<String> cFieldNames = new ArrayList(docFields.keySet());
+        for (Map.Entry<String, Double> field : docFields.entrySet()) {
+            fieldName = field.getKey();
+            cFieldNames.remove(fieldName);
+            // update running covariances
+            final double dR = delta.get(fieldName);
+            HashMap<String, Double> cFieldVals = (covariances.get(fieldName) != null) ? covariances.get(fieldName) : new HashMap<>();
+            for (String cFieldName : cFieldNames) {
+                if (cFieldVals.containsKey(cFieldName) == true) {
+                    final double newVal = cFieldVals.get(cFieldName) + 1.0/(docCount * (docCount - 1.0)) * dR * delta.get(cFieldName);
+                    cFieldVals.put(cFieldName, newVal);
+                } else {
+                    cFieldVals.put(cFieldName, 0.0);
+                }
+            }
+            if (cFieldVals.size() > 0) {
+                covariances.put(fieldName, cFieldVals);
+            }
+        }
+        dirty = true;
+    }
+
+    /**
+     * Merges the descriptive statistics of a second data set (e.g., per shard)
+     */
+    public void merge(MultiFieldStatsResults other) {
+        // update cache flag
+        dirty = true;
+        final double countA = docCount;
+        final double countB = other.docCount;
```

Comment:
Should we need a check first to make sure `other` isn't null?  I know the nulls are removed in `doReduce()` above, but thinking about future-proofing in case code gets re-arranged?